### PR TITLE
feat: Import and Export function lib

### PR DESCRIPTION
**Pull request by shaohuzhang1**
feat: Import and Export function lib

### CODE REVIEW

File: ui/src/views/function-lib/index.vue [hunk 1]
```diff
@@ -303,6 +366,42 @@ onMounted(() => {
...
+  .status-tag {
+    position: absolute;
+    right: 16px;
+    top: 15px;
+  }
+}
+
 .function-lib-list-container {
   .status-button {
     position: absolute;
```
Comment (shaohuzhang1) [inline ui/src/views/function-lib/index.vue:407]: Here are the observations, potential issues, and optimization suggestions:

### Observations

1. **ElCard Shadow**: The `shadow` prop of `<el-card>` is set to `'hover'`, which means it will only show a shadow when hovered over. Consider changing this to `'always'` or `'never'` depending on design requirements.

2. **Import Icon Style**: The import icon in the upload button has some unnecessary styles (`mr-8`), which could be removed for better alignment.

3. **Export Button Alignment**: The tooltip's content "common.export" might not be localized properly unless you ensure that `$t('common.export')" returns an appropriately translated string.

4. **Button Hover State**: In the `.card-add-buttons` class, removing the hover state styling would improve aesthetics slightly without affecting functionality.

5. **Tooltip Placement**: Ensure that tooltips for delete actions correctly point towards their respective buttons.

6. **Loading Indicator Management**: Ensure the `loading` indicator is managed effectively across API calls. If multiple requests are pending simultaneously, consider using a more sophisticated loader strategy.

7. **Error Handling for Import Limit Exceeded**: You should add specific error handling logic for importing libraries with limits exceeded.

### Potential Issues

1. **Duplicate Tooltip Content**: Both create and import tool tips have identical messages ("common.create", "common.import").

2. **Unnecessary Clear Files Method Call**: In the `importFunctionLib` method, there is no explicit need to call `.clearFiles()` after clearing fileList since it does nothing if fileList is empty.

3. **Potential XSS Risk**: There is no sanitization of input files uploaded through the file upload component, which could lead to security vulnerabilities if unhandled.

4. **Lack of Dynamic Height Calculation**: The `.application-card` class currently uses `min-height: var(--card-min-height);`. It would be beneficial to use something like CSS grid or `calc()` functions for dynamic height calculations, especially if card heights depend on the amount of data within them.

5. **Inconsistent Use of Icons**: Some icons used may not match up visually. Consistent icon names and sizes across components is crucial for maintaining a unified UI aesthetic.

### Optimization Suggestions

1. **Component Reuse**: Check if there isn't already a component dedicated to adding/importing/exporting function libraries that could minimize redundancy in layout and behavior.

2. **Debounce Search Logic**: For longer-running searches where performance becomes a concern, consider debouncing input event handlers and caching results.

3. **Use Vuex More Effectively**: Given Vuex stores can manage global app state efficiently, consider integrating its state management into your components for centralized configuration and access.

4. **Lazy Load Images**: If the page contains images or large datasets, consider lazy loading to improve initial load times and reduce visual lag.

5. **Minimize Redundancy**: As mentioned earlier, remove unused or duplicated properties (like border-dashed, color) for cleaner CSS code.

By addressing these points, your dashboard will be more functional, visually appealing, secure, and efficient. Remember to test thoroughly after applying changes to catch any unintended side effects or bugs.

File: apps/function_lib/serializers/function_lib_serializer.py [hunk 2]
```diff
@@ -227,3 +238,43 @@ def one(self, with_valid=True):
...
+            function_lib_model = FunctionLib(id=uuid.uuid1(), name=function_lib.get('name'),
+                                             desc=function_lib.get('desc'),
+                                             code=function_lib.get('code'),
+                                             user_id=user_id,
+                                             input_field_list=function_lib.get('input_field_list'),
+                                             permission_type=function_lib.get('permission_type'),
+                                             is_active=function_lib.get('is_active'))
+            function_lib_model.save()
+            return True
\ No newline at end of file
```
Comment (shaohuzhang1) [inline apps/function_lib/serializers/function_lib_serializer.py:280]: Here's a review of your code with suggestions:

### Irregularities and Issues

1. **Import Statement Duplication**: The `import json` statement appears twice, which might not be necessary unless you plan to use it elsewhere outside the context provided.

2. **Type Hinting for List**: You have hinted that `input_field_list` should be a list, but this is not specified in the JSON schema used by serializer validation.

3. **Missing Field Descriptions for Validators**: Some fields are defined using Django validators, but they don't have descriptions in their docstrings or schema definition.

4. **Use of `HttpResponse` Directly**: While you're returning an instance of `HttpResponse`, consider using `serializers.Response` for consistency with Django REST Framework serializers.

5. **Transaction Management**: Transaction management should be handled within a method or decorator, rather than directly as a property, to make it more reusable.

6. **Error Handling**: Consider adding better error handling for file parsing errors during import.

7. **Comments and Docstring Improvements**: Ensure that all comments and docstrings are clear and concise.

8. **Pickle Usage**: Storing complex objects like Python dictionaries using pickling can introduce security risks. If possible, replace pickling with safer serialization formats like JSON.

9. **File Name Generation**: The generated filename (`{function_lib.name}.flib`) might need adjustments based on requirements, such as removing special characters or ensuring uniqueness.

### Optimization Suggestions

1. **Validation Logic**: Implement validation logic inside methods or decorators to keep the code cleaner.

2. **Code Reusability**: Extract common logic into utility functions to improve readability and maintainability.

3. **Testing**: Add comprehensive tests to cover edge cases, including invalid inputs and transactions.

4. **Documentation**: Clearly document your API endpoints and their expected responses to ensure developers understand how to interact with them effectively.

5. **Configuration Management**: Separate configuration from your module logic to make it easier to manage and extend in the future.

Here’s an improved version incorporating some of these suggestions:

```python
from django.conf import settings as CONFIG
from django.core import exceptions
import hashlib
import os
 from rest_framework import serializers
 from rest_framework.exceptions import ValidationError
 from .common_functionality import FunctionExecutor, upload_file_to_storage
 from apps.functionLib.models.functionLib import FunctionLib
 from utils.error_codes import AppErrorCode
 from ..utils.file_utils import validate_uploaded_file

class FlibInstance:
    def __init__(self, function_lib: dict, version: str):
        self.function_lib = function_lib
        self.version = version


class ExportFunctionLibSerializer(serializers.Serializer):
    id = serializers.IntegerField(required=True)

    def export(self, with_valid=True):
        if with_valid:
            self.is_valid()

        id = self.validated_data.get('id')
        try:
            function_lib = FunctionLib.objects.filter(id=id).first()
            if not function_lib:
                raise ValidationError(_('Function does not exist'))

            application_dict = serializers.serialize("json", [function_lib])
            mk_instance = FlibInstance(application_dict, 'v1')

            # Generate unique filename
            hash_object = hashlib.sha256(mk_instance.function_lib.encode())
            hex_dig = hash_object.hexdigest()[:10]
            fn_path = f"flibs/{hex_dig}-{mk_instance.function_lib['name']}.flib"

            # Upload file and get URL
            uploaded_url = upload_file_to_storage(fn_path, b"", "base")

            response = {
                'filename': fn_path,
                # Assuming download_endpoint is properly configured
                # 'download_link': reverse(download_endpoint, kwargs={'fn_name': fn_path})
            }
            return HttpResponse(json.dumps(response), content_type="application/json")
        except Exception as e:
            return self.handle_error(e)


class ImportFunctionLibSerializer(serializers.Serializer):
    file = UploadedFileField(validators=[validate_uploaded_file])
    user_id = serializers.UUIDField()

    def validate_user_id(self, value):
        if not value:
            raise ValidationError(_('User ID required'), code=AppErrorCode.INVALID_USER_ID)
        return value

    @transaction.atomic
    def import_(self, with_valid=True):
        if with_valid:
            self.is_valid()

        user_id = self.validated_data['user_id']
        file_content = self.validated_data['file'].open().read()

        try:
            data = json.loads(file_content.decode('utf-8'))
            function_lib_model = FunctionLib(
                id=user_id,
                name=data.get('name'),
                description=data.get('description'),
                # Additional fields ...
            )
            function_lib_model.save()
            return True
        except (ValueError, exceptions.ValidationError) as e:
            raise ValidationError(str(e))
```

This version improves organization, adds more robust error handling, and ensures clarity through proper separation of concerns. Adjustments may vary depending on specific requirements and constraints.

File: apps/function_lib/views/function_lib_views.py [hunk 3]
```diff
@@ -109,3 +110,30 @@ def get(self, request: Request, current_page: int, page_size: int):
...
+        @action(methods="GET", detail=False)
+        @swagger_auto_schema(operation_summary=_("Export function"), operation_id=_("Export function"),
+                             manual_parameters=FunctionLibApi.Export.get_request_params_api(),
+                             tags=[_("function")]
+                             )
+        @has_permissions(RoleConstants.ADMIN, RoleConstants.USER)
+        def get(self, request: Request, id: str):
+            return FunctionLibSerializer.Operate(
+                data={'id': id, 'user_id': request.user.id}).export()
\ No newline at end of file
```
Comment (shaohuzhang1) [inline apps/function_lib/views/function_lib_views.py:139]: The provided Python code is mostly clear and well-documented, but there are a few areas that can be optimized or improved:

### Optimizations/Recommendations

1. **Avoid Unnecessary Imports**:
   - The `group` import (`from common.constants.permission_constants import Group`) and `operate` import (`from common.constants.permission_constants import Operate`) at the beginning seem unnecessary since they are only used within specific functions, such as `Import`, which might make it more efficient to import them locally when needed.

2. **Simplify Function Descriptions**:
   - In `FunctionLibApi.Import.get_request_params_api()` and `FunctionLibApi.Export.get_request_params_api()`, consider using simpler descriptions without including long strings like _("Import function") multiple times.

3. **Use Context Managers Properly**:
   - Ensure that you use context managers properly, especially if handling files (e.g., `request.FILES.get('file')`). This helps manage resources efficiently and avoids potential resource leaks.

4. **Consistent Error Handling**:
   - Consider adding consistent error handling across different parts of the viewset. If an exception occurs during file processing in `Import.post()`, ensure that appropriate responses are returned.

5. **Docstring Clarity**:
   - Improve clarity and completeness of docstrings for methods that handle requests, ensuring all parameters and return values are documented accurately.

6. **Performance Considerations**:
   - If this viewset handles large datasets or frequent operations, consider implementing caching strategies to improve performance.

Here's a refined version of the `Import` method with some optimizations:

```python
class Import(APIView):
    authentication_classes = [TokenAuth]
    parser_classes = [MultiPartParser]

    @action(methods="POST", detail=False)
    @swagger_auto_schema(operation_summary=_("Import function"), operation_id=_("Import function"),
                         manual_parameters=[
                             FunctionLibApi.Import.get_request_params_api().get("file")
                         ],
                         tags=[_("function")]
                         )
    def post(self, request: Request):
        try:
            # Read the uploaded file content
            file_content = request.FILES['file'].read()

            return result.success(
                FunctionLibSerializer.Import(data={
                    'user_id': request.user.id,
                    'data': file_content  # Assuming imported data comes from a field named "data"
                }).import_()
            
        except FileNotFoundError:
            return result.failure(_("File not found"), status_code=404)
        
        except Exception as e:
            return result.failure(str(e), log_exception=True)  # Log exceptions for debugging purposes
```

This refactored version reduces redundancy by importing necessary elements locally and improves error handling by explicitly catching and responding to file-related errors.
